static/panes/device-view.ts

```diff
@@ -0,0 +1,447 @@
+            const languageId = this.devices[this.selectedDevice].languageId;
+            this.showDeviceAsmResults(this.devices[this.selectedDevice].asm, languageId);
+        } else {
+            this.showDeviceAsmResults([{text: '<Device ' + this.selectedDevice + ' not found>'}]);
+        }
+    }
+
+    override getPaneTag() {
+        if (this._editorId) {
+            return this._compilerName + ' (Editor #' + this._editorId + ', Compiler #' + this._compilerId + ')';
+        } else {
+            return this._compilerName + ' (Tree #' + this._treeId + ', Compiler #' + this._compilerId + ')';
+        }
+    }
+
+    getDefaultPaneName() {
+        return 'Device Viewer';
+    }
+
+    override getPaneName() {
+        return this.paneName ? this.paneName : this.getDefaultPaneName() + ' ' + this.getPaneTag();
+    }
+
+    override updateTitle() {
+        this.container.setTitle(_.escape(this.getPaneName()));
+    }
+
+    showDeviceAsmResults(deviceCode: DeviceAsmCode[], languageId?: string) {
+        this.deviceCode = deviceCode;
+
+        if (!languageId) {
+            languageId = 'asm';
+        }
+
+        const model = this.deviceEditor.getModel();
+        if (model) {
+            monaco.editor.setModelLanguage(model, languageId);
+            model.setValue(deviceCode.length ? _.pluck(deviceCode, 'text').join('\n') : '<No device code>');
+        }
+
+        if (!this.awaitingInitialResults) {
+            if (this.selection) {
+                this.deviceEditor.setSelection(this.selection);
+                this.deviceEditor.revealLinesInCenter(this.selection.startLineNumber, this.selection.endLineNumber);
+            }
+            this.awaitingInitialResults = true;
+        }
+    }
+
+    onCompiler(id, compiler, options, editorId, treeId) {
+        if (id === this._compilerId) {
+            this._compilerName = compiler ? compiler.name : '';
+            this._editorId = editorId;
+            this._treeId = treeId;
+            this.updateTitle();
+            if (compiler && !compiler.supportsDeviceAsmView) {
+                this.deviceEditor.setValue('<Device output is not supported for this compiler>');
+            }
+        }
+    }
+
+    onColours(id: number, colours: Record<number, number>, scheme: string) {
+        this.lastColours = colours;
+        this.lastColourScheme = scheme;
+
+        if (id === this._compilerId) {
+            const irColours = {};
+            _.each(this.deviceCode, function (x, index) {
+                if (x.source && x.source.file == null && x.source.line > 0 && colours[x.source.line - 1]) {
+                    irColours[index] = colours[x.source.line - 1];
+                }
+            });
+            this.colours = colour.applyColours(this.deviceEditor, irColours, scheme, this.colours);
+        }
+    }
+
+    override updateState() {
+        this.container.setState(this.currentState());
+    }
+
+    currentState() {
```

Comment:
I think this can be eliminated in favor of `MonacoPane.getCurrentState` without issue

Reply:
Don't know if I did this correctly but I've replaced it

Reply:
Looks good